compiler/rustc_trait_selection/src/traits/mod.rs

```diff
@@ -525,20 +681,66 @@ fn vtable_trait_first_method_offset<'tcx>(
 ) -> usize {
     let (trait_to_be_found, trait_owning_vtable) = key;
 
-    let mut supertraits = util::supertraits(tcx, trait_owning_vtable);
+    let mut vtable_base = 0;
 
-    // For each of the non-matching predicates that
-    // we pass over, we sum up the set of number of vtable
-    // entries, so that we can compute the offset for the selected
-    // trait.
-    let vtable_base = ty::COMMON_VTABLE_ENTRIES.len()
-        + supertraits
-            .by_ref()
-            .take_while(|t| *t != trait_to_be_found)
-            .map(|t| util::count_own_vtable_entries(tcx, t))
-            .sum::<usize>();
+    let mut planner = vtable_planner(tcx, trait_owning_vtable);
+    while let GeneratorState::Yielded(segment) = Pin::new(&mut planner).resume(()) {
+        match segment {
+            VtblSegment::MetadataDSA => {
+                vtable_base += COMMON_VTABLE_ENTRIES.len();
+            }
+            VtblSegment::TraitOwnEntries { trait_ref, emit_vptr } => {
+                if trait_ref == trait_to_be_found {
+                    return vtable_base;
+                }
+                vtable_base += util::count_own_vtable_entries(tcx, trait_ref);
+                if emit_vptr {
+                    vtable_base += 1;
+                }
+            }
+        }
+    }
+
+    bug!("Failed to find info for expected trait in vtable");
+}
+
+/// Find slot offset for trait vptr within vtable entries of another trait
+#[allow(dead_code)]
```

Comment:
Is this necessary?

Reply:
This function is not yet needed in this pr, will be used in next pr. Should i remove it?

Reply:
I guess adding a FIXME to remove this line would be fine.

Reply:
Done